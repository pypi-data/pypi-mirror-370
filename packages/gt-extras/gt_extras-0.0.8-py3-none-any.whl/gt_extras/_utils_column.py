from __future__ import annotations

import warnings

from great_tables import GT
from great_tables._data_color.base import _rescale_numeric
from great_tables._locations import resolve_cols_c
from great_tables._tbl_data import SelectExpr, is_na, to_list

__all__ = [
    "_validate_and_get_single_column",
    "_scale_numeric_column",
    "_format_numeric_text",
]


def _validate_and_get_single_column(
    gt: GT,
    expr: SelectExpr,
) -> tuple[str, list]:
    """
    Validate that expr resolves to a single column and return the column name and values.

    Parameters
    ----------
    gt
        The `GT` object containing the data
    expr
        The column expression to resolve

    Returns
    -------
    tuple[str, list]
        A tuple of (column_name, column_values)

    Raises
    ------
    KeyError
        If the column is not found
    ValueError
        If multiple columns are resolved
    """
    col_names = resolve_cols_c(data=gt, expr=expr)

    if len(col_names) == 0:
        raise KeyError(f"Column '{expr}' not found in the table.")
    if len(col_names) > 1:
        raise ValueError(
            f"Expected a single column, but got multiple columns: {col_names}"
        )

    col_name = col_names[0]
    col_vals = to_list(gt._tbl_data[col_name])

    return col_name, col_vals


def _scale_numeric_column(
    data_table,
    col_name: str,
    col_vals: list,
    domain: list[float] | list[int] | None = None,
    default_domain_min_zero: bool = True,
) -> list[float]:
    """
    Process and scale a numeric column, handling NA values.

    Parameters
    ----------
    data_table
        The underlying data table
    col_name
        Name of the column (for error messages)
    col_vals
        The column values
    domain
        The domain for scaling. If None, uses a default domain, based on `default_domain_min_zero`
    default_domain_min_zero
        If true, the defsult domain will be [0, max(values)], otherwise [min(values), max(values)]
    Returns
    -------
    list[float]
        Scaled values with NAs mapped to 0, values above the domains mapped to 1,
        and values below the domain mapped to 0.

    Raises
    ------
    TypeError
        If the column is not numeric
    """
    col_vals_filtered = [x for x in col_vals if not is_na(data_table, x)]

    # Check that column has numeric data
    if len(col_vals_filtered) and all(
        isinstance(x, (int, float)) for x in col_vals_filtered
    ):
        # If `domain` is not provided, then set it to a default domain
        if domain is None:
            if default_domain_min_zero:
                domain = [0, max(col_vals_filtered)]
            else:
                domain = [min(col_vals_filtered), max(col_vals_filtered)]

        # Rescale based on the given domain
        scaled_vals = _rescale_numeric(
            df=data_table,
            vals=col_vals,
            # Alternatively we could convert the domain to floats, but I dont mind ignoring this
            domain=domain,  # type: ignore
        )
    else:
        raise TypeError(
            f"Invalid column type provided ({col_name}). Please ensure that the column is numeric."
        )

    # Map scaled values back to original positions, using 0 for NAs
    scaled_vals_fixed = []
    for orig_val, scaled_val in zip(col_vals, scaled_vals):
        if is_na(data_table, orig_val):
            scaled_vals_fixed.append(0)

        elif is_na(data_table, scaled_val):
            # consider handling by leaving the original val, and having a third color/category.

            # If original value < domain[0], set to 0; if > domain[1], set to 1
            if orig_val < min(domain):
                warnings.warn(
                    f"Value {orig_val} in column '{col_name}' is less than the domain minimum {min(domain)}. Setting to {min(domain)}.",
                    category=UserWarning,
                )
                scaled_vals_fixed.append(0)

            else:
                warnings.warn(
                    f"Value {orig_val} in column '{col_name}' is greater than the domain maximum {max(domain)}. Setting to {max(domain)}.",
                    category=UserWarning,
                )
                scaled_vals_fixed.append(1)
        else:
            scaled_vals_fixed.append(scaled_val)

    return scaled_vals_fixed


def _format_numeric_text(value: float, num_decimals: int) -> str:
    if num_decimals == 0:
        return f"{value:.0f}"
    else:
        return f"{value:.{num_decimals}f}".rstrip("0").rstrip(".")
