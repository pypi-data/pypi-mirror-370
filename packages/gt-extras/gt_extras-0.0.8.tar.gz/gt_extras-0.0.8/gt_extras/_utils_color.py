from __future__ import annotations

import warnings

from great_tables._data_color.base import (
    _get_domain_factor,
    _html_color,
    _rescale_factor,
)
from great_tables._data_color.constants import ALL_PALETTES, DEFAULT_PALETTE
from great_tables._data_color.palettes import GradientPalette
from great_tables._tbl_data import TblData

__all__ = ["_get_discrete_colors_from_palette"]


def _get_discrete_colors_from_palette(
    palette: list[str] | str | None,
    data: list,
    data_table: TblData,
) -> list[str]:
    palette = _get_palette(palette)

    # Rescale the category column for the purpose of assigning colors to each dot
    domain_factor = _get_domain_factor(df=data_table, vals=data)
    scaled_vals = _rescale_factor(
        df=data_table,
        vals=data,
        domain=domain_factor,  # type: ignore
        palette=palette,
    )

    # Create a color scale function from the palette
    color_scale_fn = GradientPalette(colors=palette)

    # Call the color scale function on the scaled categoy values to get a list of colors
    color_vals = color_scale_fn(scaled_vals)

    for i, c in enumerate(color_vals):
        if c is None:
            color_vals[i] = "transparent"
            warnings.warn(
                "A color value is None and has been coerced to 'transparent'",
                UserWarning,
            )

    # This is redundant but satisfies the type-checker
    color_vals = [c for c in color_vals if c is not None]

    return color_vals


def _get_palette(palette: list[str] | str | None) -> list[str]:
    # If palette is not provided, use a default palette
    if palette is None:
        palette = DEFAULT_PALETTE

    # Otherwise get the palette from great_tables._data_color
    elif isinstance(palette, str):
        palette = ALL_PALETTES.get(palette, [palette])

    # Standardize values in `palette` to hexadecimal color values
    palette = _html_color(colors=palette)

    return palette
