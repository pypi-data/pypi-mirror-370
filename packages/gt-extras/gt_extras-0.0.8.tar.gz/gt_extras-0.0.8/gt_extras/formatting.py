from __future__ import annotations

import re
import tempfile
import webbrowser
from pathlib import Path
from typing import TYPE_CHECKING, Literal

import narwhals.stable.v1 as nw
from great_tables import GT
from great_tables._export import _create_temp_file_server
from great_tables._gt_data import Boxhead, ColInfo
from great_tables._tbl_data import SelectExpr, is_na

from gt_extras._utils_column import _validate_and_get_single_column

__all__ = ["fmt_pct_extra", "gt_duplicate_column", "gt_two_column_layout"]


def fmt_pct_extra(
    gt: GT,
    columns: SelectExpr,
    scale: float = 100,
    threshold: float = 1.0,
    color: str = "grey",
    decimals: int = 1,
) -> GT:
    # TODO: would we want to infer the range? (as we do in gt_plt_bar_pct)
    """
    Convert to percent and show less than `1%` as `<1%` in grey.

    The `fmt_pct_extra()` function takes an existing `GT` object and formats a column of numeric
    values as percentages. Values below the specified threshold are displayed as `"<threshold%"`
    instead of their actual percentage value, and in a color of your choosing. The default behavior
    expects values in the range `[0, 1]`, use `scale` if your values are in a different range.

    Parameters
    ----------
    gt
        A `GT` object to modify.

    columns
        The columns containing numeric values to format as percentages.

    scale
        Multiplication factor to convert values to percentages. For some common cases:
        use `100` if values are decimals `(0.05 -> 5%)` (default),
        use `1` if values are already percentages `(5 -> 5%)`.

    threshold
        The percentage threshold below which values are displayed as `"<threshold%"` instead of
        their actual value. Note this refers to the scaled value, not the original, so you will
        have to provide `5` for `<5%` even if your original are in the range `[0,1]`.

    color
        The color to use for values below the threshold.

    decimals
        Number of decimal places to display for percentages.

    Returns
    -------
    GT
        A `GT` object with formatted percentage column.

    Examples
    --------
    ```{python}
    from great_tables import GT
    from great_tables.data import towny
    import gt_extras as gte

    towny_mini = towny[
        [
            "name",
            "pop_change_1996_2001_pct",
            "pop_change_2001_2006_pct",
            "pop_change_2006_2011_pct",
        ]
    ].tail(10)

    gt = (
        GT(towny_mini)
        .tab_spanner(label="Population Change", columns=[1, 2, 3])
        .cols_label(
            pop_change_1996_2001_pct="'96-'01",
            pop_change_2001_2006_pct="'01-'06",
            pop_change_2006_2011_pct="'06-'11",
        )
    )

    gt.pipe(
        gte.fmt_pct_extra,
        columns=[1, 2, 3],
        threshold=5,
        color="green",
    )
    ```
    """
    # TODO: consider how to handle negative values

    def _fmt_pct_single_val(value: float):
        if is_na(gt._tbl_data, value):
            return ""

        # Convert to percentage
        pct_value = value * scale

        if abs(pct_value) < threshold:
            return f"<span style='color:{color};'><{threshold:g}%</span>"
        else:
            return f"{pct_value:.{decimals}f}%"

    res = gt
    res = res.fmt(_fmt_pct_single_val, columns=columns)

    return res


def gt_duplicate_column(
    gt: GT,
    column: SelectExpr,
    after: str | None = None,
    append_text: str | None = "_dupe",
    dupe_name: str | None = None,
) -> GT:
    """
    Duplicate a column in a `GT` object.

    The `gt_duplicate_column()` function takes an existing `GT` object and creates a duplicate
    _(without styling)_ of the specified column. The duplicated column can be renamed using either
    `dupe_name` or by appending text to the original column name, and positioned at a
    specific location in the table.

    Parameters
    ----------
    gt
        A `GT` object to modify.

    column
        The column to duplicate. Can be a column name or index.

    after
        The column after which to place the duplicated column. If `None`, the duplicated
        column will be moved to the end of the table.

    append_text
        Text to append to the original column name for the duplicate. Only used if
        `dupe_name` is not provided. Defaults to `"_dupe"`.

    dupe_name
        The name for the duplicated column. If provided, this overrides `append_text`.

    Returns
    -------
    GT
        A `GT` object with the duplicated column added.

    Examples
    --------
    ```{python}
    from great_tables import GT
    from great_tables.data import gtcars
    import gt_extras as gte

    gtcars_mini = gtcars[["mfr", "model", "year", "hp"]].head(5)
    gt = GT(gtcars_mini)

    # Duplicate with custom name and position
    gt.pipe(
        gte.gt_duplicate_column,
        column="hp",
        after="year",
    )
    ```

    Note
    --------
    Styles generated by [`GT.tab_style()`](https://posit-dev.github.io/great-tables/reference/GT.tab_style)
    will render in the duplicated column. This may be accommodated in the future.
    """
    original_name, _ = _validate_and_get_single_column(gt, column)

    # If dupe_name is given, it overrides append_text
    append_text = append_text or "_dupe"
    if dupe_name is not None:
        new_col_name = dupe_name
    else:
        new_col_name = original_name + append_text

    if new_col_name == original_name:
        raise ValueError(
            f"The new column name '{new_col_name}' cannot be the same as the original column name '{original_name}'."
        )

    res = gt

    # get the boxhead info
    original_col_info = None
    for col_info in res._boxhead:
        if col_info.var == original_name:
            original_col_info = col_info
            break

    if TYPE_CHECKING:
        if original_col_info is None:
            raise ValueError(f"Column '{original_name}' not found in boxhead.")

    # make the new boxhead entry
    new_col_info = ColInfo(
        var=new_col_name,
        type=original_col_info.type,
        column_label=new_col_name,
        column_align=original_col_info.column_align,
        column_width=original_col_info.column_width,
    )

    # Duplicate the col in the data_table data
    original_data_table = res._tbl_data
    new_data_table = nw.to_native(
        nw.from_native(original_data_table).with_columns(
            nw.col(original_name).alias(new_col_name)
        )
    )

    # Duplicate the col in the body element of body
    new_body = res._body
    new_body.body = nw.to_native(
        nw.from_native(new_body.body).with_columns(
            nw.col(original_name).alias(new_col_name)
        )
    )

    new_boxhead_list = list(res._boxhead._d) + [new_col_info]
    new_boxhead = Boxhead(new_boxhead_list)

    res = res._replace(_tbl_data=new_data_table, _boxhead=new_boxhead, _body=new_body)

    if after is None:
        res = res.cols_move_to_end(new_col_name)
    else:
        res = res.cols_move(new_col_name, after=after)

    return res


class GTCombinedLayout:
    """
    Represents the combined layout of two `GT` objects rendered side by side.

    This class is returned by [`gt_two_column_layout()`](https://posit-dev.github.io/gt-extras/reference/gt_two_column_layout)
    and is designed for automatic rendering in Jupyter notebooks and other HTML-aware environments.

    Parameters
    ----------
    html_content
        The combined HTML content representing the two tables in a single layout.

    Methods
    -------
    _repr_html_()
        Returns the HTML content for notebook rendering.

    __str__()
        Returns the HTML content as a string.

    show()
        Displays the table in an environment of choice. Options are `"notebook"` and `"browser"`.
    """

    def __init__(self, html_content: str):
        self.html_content = html_content

    def _repr_html_(self):
        return self.html_content

    def __str__(self):
        return self.html_content

    # do a show based on code from great_tables `GT.show()`
    def show(self, target: Literal["notebook", "browser"] = "browser"):
        if target == "notebook":
            from IPython.core.display import display_html

            # https://github.com/ipython/ipython/pull/10962
            display_html(  # pyright: ignore[reportUnknownVariableType]
                self._repr_html_(), raw=True, metadata={"text/html": {"isolated": True}}
            )
        elif target == "browser":
            with tempfile.TemporaryDirectory() as tmp_dir:
                f_path = Path(tmp_dir) / "index.html"
                f_path.write_text(self._repr_html_(), encoding="utf-8")

                # create a server that closes after 1 request ----
                server = _create_temp_file_server(f_path)
                webbrowser.open(f"http://127.0.0.1:{server.server_port}/{f_path.name}")
                server.handle_request()
        else:
            raise Exception(f"Unknown target display: {target}")

    def save(self):
        raise NotImplementedError


def gt_two_column_layout(
    gt1: GT,
    gt2: GT,
    target: Literal["notebook", "browser"] | None = None,  # TODO: add save
    table_header_from: Literal[1, 2] | None = None,
) -> GTCombinedLayout:
    """
    Combine two `GT` objects into a two-column layout.

    This function takes two `GT` objects and arranges them side by side in a single HTML output.
    Optionally, the header (title and subtitle) from one of the tables can be included at the top
    of the combined layout. The output can be displayed directly in a Jupyter notebook, opened in
    a browser, or returned as HTML.

    Parameters
    ----------
    gt1
        The first table to include in the layout. This table will appear on the left side.

    gt2
        The second table to include in the layout. This table will appear on the right side.

    target
        Determines how the combined layout is displayed. Use `"notebook"` to display in a Jupyter
        notebook, `"browser"` to open in a web browser. If `None`, returns the layout object
        for automatic rendering. Currently, `"save"` is not yet implemented.

    table_header_from
        Determines which table's header (title and subtitle) to include in the combined layout.
        If set to `1`, the header from `gt1` will be used. If set to `2`, the header from `gt2`
        will be used. If `None`, no header will be included.

    Returns
    -------
    GTCombinedLayout
        A layout object that automatically renders as HTML in Jupyter notebooks and Quarto
        documents. It also has a `.show()` method, taking `"notebook"` or `"browser"` as targets.

    Examples
    --------
    ```{python}
    from great_tables import GT
    import gt_extras as gte
    import pandas as pd

    df1 = pd.DataFrame({"A": [1, 2, 3], "B": [4, 5, 6]})
    df2 = pd.DataFrame({"X": [7, 8, 9], "Y": [10, 11, 12]})

    gt1 = GT(df1).tab_header(title="Table 1", subtitle="1st Table")
    gt2 = GT(df2).tab_header(title="Table 2", subtitle="2nd Table")

    gte.gt_two_column_layout(gt1, gt2)
    ```

    Alternatively, we can combine data from the same table in a two column layout like so:

    ```{python}
    df1 = pd.DataFrame({"A": [1, 2, 3], "B": [4, 5, 6]})
    df2 = pd.DataFrame({"A": [7, 8, 9], "B": [10, 11, 12]})

    gt1 = GT(df1).tab_header(title="Joined Table")
    gt2 = GT(df2)

    gte.gt_two_column_layout(gt1, gt2, table_header_from=1)
    ```


    """

    def extract_tab_header_and_style(gt: GT) -> dict:
        """
        Extract the title, subtitle, and full style block from a GT object's HTML.
        """
        html = gt.as_raw_html()

        title = gt._heading.title if gt._heading.title else ""
        subtitle = gt._heading.subtitle if gt._heading.subtitle else ""

        # Extract the original table ID
        id_match = re.search(r'<div id="([^"]*)"', html)
        original_id = id_match.group(1) if id_match else ""

        # Extract the entire style block
        style_match = re.search(r"<style>(.*?)</style>", html, re.DOTALL)
        style_content = style_match.group(1) if style_match else ""

        # Replace the original ID with "mycombinedtable" in the CSS
        if original_id:
            style_content = style_content.replace(f"#{original_id}", "#mycombinedtable")
            style_content = style_content.replace(
                "mycombinedtable table", "mycombinedtable div"
            )

        title_td_pattern = (
            r'<td[^>]*class="[^"]*gt_title[^"]*"[^>]*>\s*'
            + re.escape(str(title))
            + r"\s*</td>"
        )
        title_match = re.search(title_td_pattern, html, re.DOTALL)

        if title_match:
            td_tag = title_match.group(0)

            # Extract class attribute
            class_match = re.search(r'class="([^"]*gt_title[^"]*)"', td_tag)
            title_class = f"gt_table {class_match.group(1)}" if class_match else ""

            # Extract style attribute
            style_match = re.search(r'style="([^"]*)"', td_tag)
            title_inline_style = style_match.group(1) if style_match else ""
        else:
            title_class = ""
            title_inline_style = ""

        subtitle_class = ""

        subtitle_td_pattern = (
            r'<td[^>]*class="[^"]*gt_subtitle[^"]*"[^>]*>\s*'
            + re.escape(str(subtitle))
            + r"\s*</td>"
        )
        subtitle_match = re.search(subtitle_td_pattern, html, re.DOTALL)

        if subtitle_match:
            td_tag = subtitle_match.group(0)

            # Extract class attribute
            class_match = re.search(r'class="([^"]*gt_subtitle[^"]*)"', td_tag)

            if class_match:
                # Exclude gt_bottom_border from class string if present
                class_str = class_match.group(1)
                class_str = " ".join(
                    c for c in class_str.split() if c != "gt_bottom_border"
                )
                subtitle_class = f"gt_table {class_str}"

            # Extract style attribute
            style_match = re.search(r'style="([^"]*)"', td_tag)
            subtitle_inline_style = style_match.group(1) if style_match else ""
        else:
            subtitle_inline_style = ""

        return {
            "title": title,
            "subtitle": subtitle,
            "title_class": title_class,
            "subtitle_class": subtitle_class,
            "title_style": title_inline_style,
            "subtitle_style": subtitle_inline_style,
            "style": style_content,
        }

    # Extract header data if specified
    if table_header_from == 1:
        header_data = extract_tab_header_and_style(gt1)
    elif table_header_from == 2:
        header_data = extract_tab_header_and_style(gt2)
    else:
        header_data = None

    # Build the header HTML if header data is available
    if header_data and (header_data["title"] or header_data["subtitle"]):
        header_html = f"""
        <style>
        {header_data["style"]}
        </style>
        """

        # Add title if it exists
        if header_data["title"]:
            header_html += f"""
            <div class="{header_data["title_class"]}"
                style="width:100%; {header_data["title_style"]}"
            >
                {header_data["title"]}
            </div>
            """

        # Add subtitle if it exists
        if header_data["subtitle"]:
            header_html += f"""
            <div class="{header_data["subtitle_class"]}"
                style="width:100%; {header_data["subtitle_style"]}">
                {header_data["subtitle"]}
            </div>
            """

        # Remove headers from the two tables
        gt1_no_header = gt1.tab_header(title=None, subtitle=None)  # type: ignore
        table_1_html = gt1_no_header.as_raw_html()
        gt2_no_header = gt2.tab_header(title=None, subtitle=None)  # type: ignore
        table_2_html = gt2_no_header.as_raw_html()

    else:
        header_html = ""
        table_1_html = gt1.as_raw_html()
        table_2_html = gt2.as_raw_html()

    double_table_html = f"""
    <div style="display: flex; justify-content: center; width: 100%;">
        <div id="mycombinedtable" style="display: inline-block; width: auto;">
            {header_html}
            <div style="overflow: auto; white-space: nowrap;">
                <div style="display: inline-block; margin-right: 1em;">
                    {table_1_html}
                </div>
                <div style="display: inline-block;">
                    {table_2_html}
                </div>
            </div>
        </div>
    </div>
    """

    double_table = GTCombinedLayout(double_table_html)

    if target:
        if target == "save":
            raise NotImplementedError(
                "At the moment, only notebook and browser display options are available."
            )
        else:
            double_table.show(target)

    return GTCombinedLayout(double_table_html)
