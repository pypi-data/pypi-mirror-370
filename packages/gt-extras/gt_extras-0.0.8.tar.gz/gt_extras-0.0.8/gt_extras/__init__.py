# Import objects from the module
from .colors import (
    gt_color_box,
    gt_data_color_by_group,
    gt_highlight_cols,
    gt_highlight_rows,
    gt_hulk_col_numeric,
)
from .formatting import fmt_pct_extra, gt_duplicate_column, gt_two_column_layout
from .html import gt_merge_stack, with_hyperlink, with_tooltip
from .icons import fa_icon_repeat, gt_fa_rank_change, gt_fa_rating
from .images import add_text_img, gt_fmt_img_circle, img_header
from .plotting import (
    gt_plt_bar,
    gt_plt_bar_pct,
    gt_plt_bar_stack,
    gt_plt_bullet,
    gt_plt_conf_int,
    gt_plt_donut,
    gt_plt_dot,
    gt_plt_dumbbell,
    gt_plt_winloss,
)
from .styling import gt_add_divider
from .summary import gt_plt_summary
from .themes import (
    gt_theme_538,
    gt_theme_dark,
    gt_theme_dot_matrix,
    gt_theme_espn,
    gt_theme_excel,
    gt_theme_guardian,
    gt_theme_nytimes,
    gt_theme_pff,
)

__all__ = [
    "gt_theme_538",
    "gt_theme_espn",
    "gt_theme_nytimes",
    "gt_theme_guardian",
    "gt_theme_excel",
    "gt_theme_dot_matrix",
    "gt_theme_dark",
    "gt_theme_pff",
    "gt_data_color_by_group",
    "gt_highlight_cols",
    "gt_highlight_rows",
    "gt_hulk_col_numeric",
    "gt_color_box",
    "fa_icon_repeat",
    "gt_fa_rank_change",
    "gt_fa_rating",
    "gt_plt_bar",
    "gt_plt_bar_pct",
    "gt_plt_bullet",
    "gt_plt_dot",
    "gt_plt_conf_int",
    "gt_plt_dumbbell",
    "gt_plt_donut",
    "gt_plt_winloss",
    "gt_plt_bar_stack",
    "with_hyperlink",
    "with_tooltip",
    "gt_merge_stack",
    "fmt_pct_extra",
    "gt_duplicate_column",
    "gt_two_column_layout",
    "add_text_img",
    "img_header",
    "gt_fmt_img_circle",
    "gt_add_divider",
    "gt_plt_summary",
]
