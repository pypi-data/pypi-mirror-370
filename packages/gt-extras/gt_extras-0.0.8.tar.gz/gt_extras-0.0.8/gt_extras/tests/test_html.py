import numpy as np
import pandas as pd
import pytest
from great_tables import GT

from gt_extras import gt_merge_stack, with_hyperlink, with_tooltip
from gt_extras.tests.conftest import assert_rendered_body


def test_with_hyperlink_basic():
    result = with_hyperlink("Google", "https://google.com")
    expected = '<a href="https://google.com" target="_blank">Google</a>'
    assert result == expected


def test_with_hyperlink_new_tab_false():
    result = with_hyperlink("Google", "https://google.com", new_tab=False)
    expected = '<a href="https://google.com" target="_self">Google</a>'
    assert result == expected


def test_with_hyperlink_new_tab_true():
    result = with_hyperlink("GitHub", "https://github.com", new_tab=True)
    expected = '<a href="https://github.com" target="_blank">GitHub</a>'
    assert result == expected


def test_with_hyperlink_empty_text():
    result = with_hyperlink("", "https://example.com")
    expected = '<a href="https://example.com" target="_blank"></a>'
    assert result == expected


def test_with_hyperlink_in_table():
    df = pd.DataFrame(
        {
            "Name": ["Google", "GitHub"],
            "Link": [
                with_hyperlink("Visit Google", "https://google.com"),
                with_hyperlink("View GitHub", "https://github.com", new_tab=False),
            ],
        }
    )

    gt = GT(df)
    html_output = gt.as_raw_html()

    assert (
        '<a href="https://google.com" target="_blank">Visit Google</a>' in html_output
    )
    assert "https://github.com" in html_output
    assert 'target="_blank"' in html_output
    assert 'target="_self"' in html_output


def test_with_tooltip_basic():
    result = with_tooltip("1", "Number One")
    expected = '<abbr style="cursor: help; text-decoration: underline; text-decoration-style: dotted; color: blue; " title="Number One">1</abbr>'
    assert result == expected


def test_with_tooltip_underline_style():
    result = with_tooltip("1", "Number One", text_decoration_style="solid")
    expected = '<abbr style="cursor: help; text-decoration: underline; text-decoration-style: solid; color: blue; " title="Number One">1</abbr>'
    assert result == expected


def test_with_tooltip_underline_fail():
    with pytest.raises(ValueError):
        with_tooltip("1", "Number One", text_decoration_style="underline")  # type: ignore


def test_with_tooltip_None_color_fail():
    with pytest.raises(ValueError):
        with_tooltip("1", "Number One", color=None)  # type: ignore


def test_with_tooltip_underline_style_none():
    result = with_tooltip("1", "Number One", text_decoration_style="none")
    expected = '<abbr style="cursor: help; text-decoration: none; color: blue; " title="Number One">1</abbr>'
    assert result == expected


def test_with_tooltip_color_none_pass():
    result = with_tooltip("1", "Number One", color="none")
    expected = '<abbr style="cursor: help; text-decoration: underline; text-decoration-style: dotted; " title="Number One">1</abbr>'
    assert result == expected


def test_with_tooltip_custom_color():
    result = with_tooltip("1", "Number One", color="red")
    expected = '<abbr style="cursor: help; text-decoration: underline; text-decoration-style: dotted; color: red; " title="Number One">1</abbr>'
    assert result == expected


def test_with_tooltip_in_table():
    df = pd.DataFrame(
        {
            "Number": ["1", "2"],
            "Description": [
                with_tooltip("1", "Number One"),
                with_tooltip(
                    "2", "Number Two", text_decoration_style="solid", color="red"
                ),
            ],
        }
    )

    html_output = GT(df).as_raw_html()

    assert 'title="Number One"' in html_output
    assert 'title="Number Two"' in html_output
    assert "cursor: help" in html_output
    assert "text-decoration-style: dotted" in html_output
    assert "text-decoration-style: solid" in html_output
    assert "color: blue" in html_output
    assert "color: red" in html_output


@pytest.fixture
def sample_gt():
    sample_df = pd.DataFrame(
        {
            "col1": ["AA", "BB", "CC", np.nan],
            "col2": ["XX", "YY", "ZZ", None],
            "col3": ["1", "2", "3", "4"],
        }
    )
    return GT(sample_df)


def test_gt_merge_stack_basic(sample_gt):
    gt = gt_merge_stack(
        sample_gt,
        col1="col1",
        col2="col2",
    )
    html = gt.as_raw_html()

    assert html.count("AA") == 1
    assert html.count("BB") == 1
    assert html.count("CC") == 1
    assert html.count("XX") == 1
    assert html.count("YY") == 1
    assert html.count("ZZ") == 1

    assert html.count("</span>") == 8
    assert html.count("color:grey;") == 4


def test_gt_merge_stack_snap(snapshot):
    df = pd.DataFrame({"col1": ["A", "B", "C"], "col2": ["X", 1, None]})
    gt = GT(df)
    res = gt_merge_stack(gt, col1="col1", col2="col2")
    assert_rendered_body(snapshot, gt=res)


def test_gt_merge_stack_empty_columns():
    sample_df = pd.DataFrame(
        {
            "col1": [None, None, np.nan],
            "col2": [None, np.nan, np.nan],
        }
    )
    gt = GT(sample_df)
    gt = gt_merge_stack(
        gt,
        col1="col1",
        col2="col2",
    )
    html = gt.as_raw_html()

    assert "None" not in html
    assert "nan" not in html


def test_gt_merge_stack_mixed_values():
    sample_df = pd.DataFrame(
        {
            "col1": ["AA", 2, None, np.nan, "EE"],
            "col2": [None, "YY", 3.5, np.nan, "ZZ"],
        }
    )
    gt = GT(sample_df)
    gt = gt_merge_stack(
        gt,
        col1="col1",
        col2="col2",
    )
    html = gt.as_raw_html()

    assert "AA" in html
    assert "2" in html
    assert "EE" in html
    assert "YY" in html
    assert "3.5" in html
    assert "ZZ" in html

    assert "None" not in html
    assert "nan" not in html


def test_gt_merge_stack_custom_styles(sample_gt):
    gt = gt_merge_stack(
        sample_gt,
        col1="col1",
        col2="col2",
        font_size_main=18,
        font_size_secondary=12,
        font_weight_main="lighter",
        font_weight_secondary="bold",
        color_main="blue",
        color_secondary="red",
        small_caps=False,
    )
    html = gt.as_raw_html()

    assert html.count("font-size:18px;") == 4
    assert html.count("font-size:12px;") == 4
    assert html.count("font-weight:lighter;") == 4
    assert html.count("font-weight:bold;") == 4
    assert html.count("color:blue;") == 4
    assert html.count("color:red;") == 4
    assert html.count("small-caps") == 0
