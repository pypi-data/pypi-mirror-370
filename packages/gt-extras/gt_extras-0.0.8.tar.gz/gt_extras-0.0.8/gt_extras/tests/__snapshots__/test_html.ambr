# serializer version: 1
# name: test_gt_merge_stack_snap
  '''
  <tbody class="gt_table_body">
    <tr>
      <td class="gt_row gt_left"><div>
              <div style="line-height:14px;">
              <span style="
                  font-weight:bold;
                  font-variant:small-caps;
                  color:black;
                  font-size:14px;
              ">
                  A
              </span>
          </div>
              <div style="line-height:10px;">
              <span style="
                  font-weight:normal;
                  color:grey;
                  font-size:10px;
              ">
                  X
              </span>
          </div>
          </div></td>
    </tr>
    <tr>
      <td class="gt_row gt_left"><div>
              <div style="line-height:14px;">
              <span style="
                  font-weight:bold;
                  font-variant:small-caps;
                  color:black;
                  font-size:14px;
              ">
                  B
              </span>
          </div>
              <div style="line-height:10px;">
              <span style="
                  font-weight:normal;
                  color:grey;
                  font-size:10px;
              ">
                  1
              </span>
          </div>
          </div></td>
    </tr>
    <tr>
      <td class="gt_row gt_left"><div>
              <div style="line-height:14px;">
              <span style="
                  font-weight:bold;
                  font-variant:small-caps;
                  color:black;
                  font-size:14px;
              ">
                  C
              </span>
          </div>
              <div style="line-height:10px;">
              <span style="
                  font-weight:normal;
                  color:grey;
                  font-size:10px;
              ">
                  
              </span>
          </div>
          </div></td>
    </tr>
  </tbody>
  '''
# ---
